domain/modelconfig/validators/validators_test.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package validators
+
+import (
+	"github.com/juju/errors"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/environs/config"
+)
+
+type dummySecretBackendProviderFunc func(string) (bool, error)
+
+type dummySpaceProviderFunc func(string) (bool, error)
+
+type validatorsSuite struct{}
+
+var _ = gc.Suite(&validatorsSuite{})
+
+func (d dummySecretBackendProviderFunc) HasSecretsBackend(s string) (bool, error) {
+	return d(s)
+}
+
+func (d dummySpaceProviderFunc) HasSpace(s string) (bool, error) {
+	return d(s)
+}
+
+func (_ *validatorsSuite) TestCharmhubURLChange(c *gc.C) {
+	oldCfg, err := config.New(config.NoDefaults, map[string]any{
+		"name":         "wallyworld",
+		"uuid":         "a677bdfd-3c96-46b2-912f-38e25faceaf7",
+		"type":         "sometype",
+		"charmhub-url": "https://charmhub.example.com",
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	newCfg, err := config.New(config.NoDefaults, map[string]any{
+		"name":         "wallyworld",
+		"uuid":         "a677bdfd-3c96-46b2-912f-38e25faceaf7",
+		"type":         "sometype",
+		"charmhub-url": "https://charmhub1.example.com",
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	var validationError *config.ValidationError
+	_, err = CharmhubURLChange()(newCfg, oldCfg)
+	c.Assert(errors.As(err, &validationError), jc.IsTrue)
+	c.Assert(validationError.InvalidAttrs, gc.DeepEquals, []string{"charmhub-url"})
+}
+
+func (_ *validatorsSuite) TestCharmhubURLNoChange(c *gc.C) {
+	oldCfg, err := config.New(config.NoDefaults, map[string]any{
+		"name":         "wallyworld",
+		"uuid":         "a677bdfd-3c96-46b2-912f-38e25faceaf7",
+		"type":         "sometype",
+		"charmhub-url": "https://charmhub.example.com",
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	newCfg, err := config.New(config.NoDefaults, map[string]any{
+		"name":         "wallyworld",
+		"uuid":         "a677bdfd-3c96-46b2-912f-38e25faceaf7",
+		"type":         "sometype",
+		"charmhub-url": "https://charmhub.example.com",
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	_, err = CharmhubURLChange()(newCfg, oldCfg)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (_ *validatorsSuite) TestSpaceCheckerFound(c *gc.C) {
+	provider := dummySpaceProviderFunc(func(s string) (bool, error) {
+		c.Assert(s, gc.Equals, "foobar")
+		return true, nil
+	})
+
+	oldCfg, err := config.New(config.NoDefaults, map[string]any{
+		"name":          "wallyworld",
+		"uuid":          "a677bdfd-3c96-46b2-912f-38e25faceaf7",
```

Comment:
We have a ModelTag in testing package we tend to use in all the tests to get a model uuid to use.